clang/lib/Sema/SemaDecl.cpp

```diff
@@ -6729,7 +6729,10 @@ NamedDecl *Sema::ActOnVariableDeclarator(
         // proper storage class for other tools to use even if we're not going
         // to emit any code for it.
         NewVD->setTSCSpec(TSCS);
-      } else
+      } else if (!getLangOpts().SYCLIsDevice)
```

Comment:
In order to upstream this patch, there must be clear justification for deviation from other offloading models implementation.
Please, clarify new `if -statement` added instead of adjusting `if-statement` at line 6723.

Reply:
May I go ahead with this change?

Reply:
> In the SYCL case, we are not marking the data for the device,

This sounds wrong to me. We must mark the data for the device too.
I think @Fznamznon already fixed this.

SYCL shouldn't be different from OpenMP in how we diagnose TLS support.

Reply:
Is there a SYCL equivalent of OpenMP's:
#pragma omp declare target
int i;  // mapped to device
#pragma omp end declare target
?

In the header file that I am trying to get to compile cleanly, all I see is an unguarded
extern __thread void *foo;

Isn't this conceptually different?

Reply:
SYCL compiler adds `sycl_device` attribute to symbols reachable from SYCL kernels.

I.e. if `foo` in your example is used in the device code, SYCL compiler will mark `foo` with `sycl_device` attribute.
We must report an error if both of them (`__thread` and `sycl_device`) are applied at the same time.
Should be no different from my POV.

Reply:
> SYCL compiler adds `sycl_device` attribute to symbols reachable from SYCL kernels.
> 
> I.e. if `foo` in your example is used in the device code, SYCL compiler will mark `foo` with `sycl_device` attribute.
> We must report an error if both of them (`__thread` and `sycl_device`) are applied at the same time.
> Should be no different from my POV.

Alexey, I don't see this at all.  If I modify my code such that hard errors are disabled and such that we enter the block at 6795 (i.e. the if condition that checks if emitTLSUnsupportedError is true), there are no other attributes on the TLS variables.

Reply:
Right. They are added later, once we parsed the kernels and find that these declarations are used in the kernels. Until we know that, we should report any errors.
Please, take a look at implementation of other diagnostics for SYCL device code.

Reply:
And this is what I do as well for this test case.   At the point of use, I diagnose the thread-local variable.  Please see my changes in SemaSYCL.cpp.

Reply:
So, what is the principal difference between the SYCL and OpenMP, which doesn't allow us to modify statement at 6723?

Reply:
With OpenMP, the declaration with the pragma directives around it add the attribute to the VD, so the modification at 6723 is caught at 6795.

With SYCL, the declaration does not have the attribute attached to the VD at that time, so if I change it similarly at 6723, I can't detect it at 6795.

Reply:
So the intent is to check nothing at this stage. Right?
Wouldn't it be more clear to add change 6722?

`!Context.getTargetInfo().isTLSSupported() || getLangOpts().SYCLIsDevice)`

Also, I suggest re-using existing `diag::err_thread_unsupported` diagnostics.

Reply:
Not sure if clearer to change it to:
`!Context.getTargetInfo().isTLSSupported() && !getLangOpts().SYCLIsDevice)`
but I can do that.

I will change the diagnostics to use `diag::err_thread_unsupported`.

Reply:
For me I'd find it clearer if there was an if block for SYCLIsDevice with a comment explaining the difference:

      if (getLangOpts().CUDA || getLangOpts().OpenMPIsDevice) {
        ...
      } else if (getLangOpts().SYCLIsDevice) {
        // Postponed until <whatever>
        NewVD->setTSCSpec(TSCS);
      } else
        Diag(D.getDeclSpec().getThreadStorageClassSpecLoc(),
             diag::err_thread_unsupported);

But I don't have a big issue with it as is.

Reply:
> For me I'd find it clearer if there was an if block for SYCLIsDevice with a comment explaining the difference:

      if (getLangOpts().CUDA || getLangOpts().OpenMPIsDevice) {
        ...
      } else if (getLangOpts().SYCLIsDevice) {
        // Postponed until <whatever>
        NewVD->setTSCSpec(TSCS);
      } else
        Diag(D.getDeclSpec().getThreadStorageClassSpecLoc(),
             diag::err_thread_unsupported);

Make sense to me.

Reply:
Thanks @mikerice1969 and @bader, I have made this change as well as the diagnostic change.